examples/fluids/src/setuplibceed.c

```diff
@@ -213,33 +210,230 @@ PetscErrorCode SetupBCQFunctions(Ceed ceed, PetscInt dim_sur, PetscInt num_comp_
+// -----------------------------------------------------------------------------
+PetscErrorCode BasisCreateFromTabulation(Ceed ceed, DM dm, DMLabel domain_label, PetscInt label_value, PetscInt height, PetscInt face, PetscFE fe,
+                                         PetscTabulation basis_tabulation, PetscQuadrature quadrature, CeedBasis *basis) {
+  PetscInt           first_point;
+  PetscInt           ids[1] = {label_value};
+  DMLabel            depth_label;
+  DMPolytopeType     cell_type;
+  CeedElemTopology   elem_topo;
+  PetscScalar       *q_points, *interp, *grad;
+  const PetscScalar *q_weights;
+  PetscDualSpace     dual_space;
+  PetscInt           num_dual_basis_vectors;
+  PetscInt           dim, num_comp, P, Q;
+
+  PetscFunctionBeginUser;
+
+  // General basis information
+  PetscCall(PetscFEGetSpatialDimension(fe, &dim));
+  PetscCall(PetscFEGetNumComponents(fe, &num_comp));
+  PetscCall(PetscFEGetDualSpace(fe, &dual_space));
+  PetscCall(PetscDualSpaceGetDimension(dual_space, &num_dual_basis_vectors));
+  P = num_dual_basis_vectors / num_comp;
+
+  // Use depth label if no domain label present
+  if (!domain_label) {
+    PetscInt depth;
+
+    PetscCall(DMPlexGetDepth(dm, &depth));
+    PetscCall(DMPlexGetDepthLabel(dm, &depth_label));
+    ids[0] = depth - height;
+  }
+
+  // Get cell interp, grad, and quadrature data
+  PetscCall(DMGetFirstLabeledPoint(dm, dm, domain_label ? domain_label : depth_label, 1, ids, height, &first_point, NULL));
+  PetscCall(DMPlexGetCellType(dm, first_point, &cell_type));
+  elem_topo = ElemTopologyP2C(cell_type);
+  if (!elem_topo) SETERRQ(PetscObjectComm((PetscObject)dm), PETSC_ERR_SUP, "DMPlex topology not supported");
+  {
+    size_t             q_points_size;
+    const PetscScalar *q_points_petsc;
+    PetscInt           q_dim;
+
+    PetscCall(PetscQuadratureGetData(quadrature, &q_dim, NULL, &Q, &q_points_petsc, &q_weights));
+    q_points_size = Q * dim * sizeof(CeedScalar);
+    PetscCall(PetscCalloc(q_points_size, &q_points));
+    for (PetscInt q = 0; q < Q; q++) {
+      for (PetscInt d = 0; d < q_dim; d++) q_points[q * dim + d] = q_points_petsc[q * q_dim + d];
+    }
+  }
+
+  {  // Convert to libCEED orientation
+    PetscBool       is_simplex  = PETSC_FALSE;
+    IS              permutation = NULL;
+    const PetscInt *permutation_indices;
+
+    PetscCall(DMPlexIsSimplex(dm, &is_simplex));
+    if (!is_simplex) {
+      PetscSection section;
+
+      // -- Get permutation
+      PetscCall(DMGetLocalSection(dm, &section));
+      PetscCall(PetscSectionGetClosurePermutation(section, (PetscObject)dm, dim, num_comp * P, &permutation));
+      PetscCall(ISGetIndices(permutation, &permutation_indices));
+    }
+
+    // -- Copy interp, grad matrices
+    PetscCall(PetscCalloc(P * Q * sizeof(CeedScalar), &interp));
+    PetscCall(PetscCalloc(P * Q * dim * sizeof(CeedScalar), &grad));
+    const CeedInt c = 0;
+    for (CeedInt q = 0; q < Q; q++) {
```

Comment:
What is the point of `c` here if it's always 0?

Reply:
That is a good point. I have no idea. That code was taken from Ratel: https://gitlab.com/micromorph/ratel/-/blob/main/src/internal/ratel-dm.c?ref_type=heads#L292-302

@jeremylt any ideas?

Reply:
See the source of the tabulation here:
https://petsc.org/release//manualpages/FE/PetscFEGetCellTabulation/

PETSc's tabulation contains separate rows/cols for each component, but libCEED handles that with tensor contractions, so we only need one (WLOG the first) component's rows/columns.

Reply:
This may need to be revisited if we start using different FE spaces